### ISSUE TITLE
Using protobuf for netcode?

### ISSUE THREAD

**Issue by necessarily-equal** [DaemonEngine/Daemon#762]
First, a bit of context. The current netcode sends raw udp datagrams (with Huffman encoding for entities). It does a lot of manual things like sequence numbers, "reliable commands", "reliable sequences". Understanding how it all is supposed to work is quite terrible IMO. And modifying it is worse. One has to write the code twice, once for sending and once for receiving. It is easy to get wrong, and I must say I have some doubts on the resilience against untrusted input. For example I wouldn't be surprised if a griver could do a denial of service against a server.

Now about what I'm attempting to do at the moment. I'm a bit stuck for https://github.com/DaemonEngine/Daemon/pull/744. What I'm trying to do in that PR is to have the information about the `struct entityState_t` stored in the game and not the engine. Basically this is done with an array that says "at offset 48 you will find a unsigned integer of width 10 that represents the model". And the engine then sends only 10 bytes from that offset (over raw udp with Huffman encoding) when the value changes.

It's great, it's network efficient, it's not json! From https://github.com/DaemonEngine/Daemon/blob/master/src/engine/server/sv_client.cpp#L335_L406, one can notice that there are two important things sent at the beginning of a game. 1. so-called "config strings" and 2. "baseline entities". Config strings will tell information on the server, such as maps, or versions of the game code. So if the client has some outdated game code, in theory, the client should download the new version of the gamecode and then [start again](https://github.com/DaemonEngine/Daemon/blob/master/src/engine/server/sv_client.cpp#L498_L508).

The problem I'm facing is that the message that tells the client that the client and the the server have different game code is the same message that sends those "baseline entities". The client and the server gamelogic may have different `entityState_t` definitions which would then be sent quite differently over the network. This essentially means I'm receiving an message that I don't know how to parse
but I have to read that message correctly anyway, as otherwise I get some desynchronisation in the protocol and it blows up.

I've seen @illwieckz suggesting me needed something else for the networking code, that didn't need a PhD to be able to work with it, and I've seen @DolceTriade complaining about configstring too (https://github.com/Unvanquished/Unvanquished/issues/2097), which are related but a lesser pain IMO.

Protocol buffers (protobuf) are documented have several implementations, are maintained by google and would likely be maintained by other even if google stopped caring about it (which shouldn't happen); it's still pretty efficient in message size (more than json and binary json-like formats);, it allows for keeping backward and forward compat much more easily. Using protobuf we could keep deltaents support like we are doing currently very easily too.

The downsides I can think of are that:
1. this adds a compile-time step
2. this adds a dependency
3. the fields have a tag, which means we send ~one more byte for each field

Additionaly, we no longer have weird things like "integers of size 10" and instead it is use variable length integers that will either take 1 byte or 2 bytes for a 10 byte int. This isn't really either a downside or something advantageous, it's simply different. And lastly, the message preparation would take place in gamelogic. That's not really a downside, but still something to have in mind.

Thoughts on this?

**Comment by DolceTriade** [comment]
I don't think protobufs are great for netcode. They don't compress well, and some protobufs don't serialize deterministically (like maps) which will cause issues when trying to encode them. I think the approach of moving the netcode definitions to gamelogic is probably enough to appease me.

As for info strings...eh. I don' t like infostrings and think they're obsolete and think using a subset of JSON is probably better, but its not a problem that particularly interests me because infostrings generally work.

**Comment by necessarily-equal** [comment]
For the record, I don't think our current binary format compresses well either. I don't think we are going to use maps, but I don't see what you mean about "serialize deterministically", do you mean the ordering might change? In which case this isn't a problem in our case? (it might be a problem for a diff or a db, but we don't do these)

**Comment by slipher** [comment]
I think the only problem this would really solve is automatically generating a struct and allowing reflection on it, so that we wouldn't have to keep the table with `PSF(blah)` up to date. Everything else is a disadvantage and there is a lot of impedance mismatch with what we currently do.

- IIRC fixed-length arrays like `float origin[3]` are not supported
- It would be an attractive nuisance for people to add e.g. `string` fields which could be too big
- Although *merging* a delta happens to be a supported operation, I don't think there is support for creating one. We'd have to do that using the reflection ("descriptors").
- The generated structs have a presence bit (`has_foo()`) for each field which we don't want
- Uglier accessors like `es.set_foo(5)` instead of `es.foo = 5`

**Comment by necessarily-equal** [comment]
> I was talking about our diff'ing netcode which only sends the changes between two different snapshots. For protobuf, would we send a fullsnapshot each time?

@DolceTriade: No we wouldn't, I think the easiest is to have almost each field optional (except entityNum I guess), and send only those which have been modified. This isn't more complicated to implement than what the current code does.

> IIRC fixed-length arrays like float origin[3] are not supported

@slipher (for this and for the following quotes): Indeed. I think we should use packed for these, and maybe even put origin and angles in the same field to save a bit of space. And maybe origin2 and angles2 too.

> It would be an attractive nuisance for people to add e.g. string fields which could be too big

I mean people can already use MSG_WriteString(). And to be honest, if we manage to make it easier for people to do stuff, I count that as a plus.

> Although merging a delta happens to be a supported operation, I don't think there is support for creating one. We'd have to do that using the reflection ("descriptors").

See my answer to DolceTriade at the beginning

> The generated structs have a presence bit (has_foo()) for each field which we don't want

I mean, we will just not use it then? For the record I don't want the protobuf struct to replace playerState_t directly, just the netcode format.

> Uglier accessors like es.set_foo(5) instead of es.foo = 5

Well, I mean, yes, but it's still better than fiddling with MSG_WriteLong() for these. See also: « For the record I don't want the protobuf struct to replace playerState_t directly, just the netcode format. »

**Comment by necessarily-equal** [comment]
Another advantage of using protobuf for the network message format would be that we would document the network protocol. Here's an example of what it might look like:

```proto
syntax = "proto3";

message ServerCmd {
  uint32 sequence = 1;
  string command = 2;
}

message ServerSnapshot {
  byte ents = 1; // handled by gamelogic
}

// [define the other message types here]

message ServerMsg {
  uint32 reliable_acknowledge = 1;
  fixed32 crc32 = 2; // (we don't have that yet AFAIK)

  oneof content { // a message contains either of those
    ServerCmdMsg sv_cmd = 3;
    ServerSnapshotMsg sv_snap = 4;
    // etc
  }
}

message ServerProtocol {
  repeated ServerMsg msgs = 1;
}
```

I find that it makes it much clearer what fields compose a message and what fields are expected to be part of each message. It also makes it easier to document what each field is about.

---

By the way, another possible advantage of protobuf: wireshark [can decode protobuf UDP messages](https://wiki.wireshark.org/Protobuf). Not sure we would use it and we would need to turn the whole protocol into protobuf first; but hey, that would be quite practical.

**Comment by slipher** [comment]
At first I took your proposal to be about using protobuf for entityState_t/playerState_t, which seemed unpromising. Using it for everything *except* those is interesting. We could keep using the old table-based approach for playerState_t to parse the internal byte array. I agree that the message parsing code is likely to contain vulnerabilities and it would nice to have a self-documenting protocol. The overhead should be small since the entity/player states comprise the bulk of data (while seldom-used messages comprise the bulk of protocol complexity).

**Comment by DolceTriade** [comment]
fwiw, proto3 removes the `required` and `optional` tags so we don't get the `msg.has_foo` things since proto3 doesn't support that.

The build time dependency on `protoc` is annoying, but workable. Integration into cmake is also workable.

> the fields have a tag, which means we send ~one more byte for each field
It's a varint so it depends on how many fields we have per message 😄  If we have only a small number of fields in each message, the overhead will be smaller.

Like slipher, I missed the fact that this was going to keep the current playerState_t and entityState_t code as-is, which makes it more viable.

Ultimately, I agree, it would be a net-plus, but its not so much a net plus that I would prioritize it myself 🤷

**Comment by necessarily-equal** [comment]
I may work on it. On pain point area for me would be CMake, and testing on not-linux. Would it be fine to use FindProtobuf and rely on people installing it, or is it better to include protobuf as a submodule and try to build it?

**Comment by slipher** [comment]
You cannot rely on people installing it (particularly on Windows).

If you build it, then of course you have to use a separate CMake invocation so that cross-compile builds work :(

The other option is to ship proto compiler binaries in external_deps, similar to how we ship the NaCl compiler. But you can't follow NaCl's example, as it doesn't work with cross-compiles. You would have to add a new set of packages which are based on the host platform rather than the target platform.

**Comment by DolceTriade** [comment]
We could just commit the generated code. I don't anticipate the protos changing much and people changing the protos could install protoc themselves... 🤷

**Comment by slipher** [comment]
Yeah committing the generated code seems better, as long as it's not platform-dependent.

Having it in external_deps actually might not work because there is a C++ runtime component and you can't build C++ libraries that are compatible with different compilers.

We could consider using Cap'n Proto instead which has a somewhat smaller repo.

**Comment by DolceTriade** [comment]
There's also flatbuffers which is more for games: https://google.github.io/flatbuffers/

**Comment by illwieckz** [comment]
That flatbuffers thing looks interesting!

**Comment by slipher** [comment]
What's the advantage of Flatbuffers? According to its own benchmarks the compressed wire format uses 26% more space than Protobuf. I see this as our main performance criterion.

**Comment by DolceTriade** [comment]
Flatbuffers shares a similar design as Capn Proto in that flat buffers structs can be used without [de]serialization. Protobufs have a cost in serialization and deserialization.
